In [17]:
# Imports section
import pandas as pd 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import make_pipeline

## Part 1. Loading the dataset

In [18]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
df = pd.read_csv("science_data_large.csv")
print(df.head(15))
print(df.info())
print(df.describe())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [19]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = df[['Temperature °C', 'Mols KCL']]  
y = df['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## Part 3. Perform a Linear Regression

In [20]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample = [[750, 840]] 
predicted_size = model.predict(sample)
print(f"Predicted size for sample {sample}: {predicted_size[0]}")
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print(f"Model R² Score: {score:.4f}")
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_
print(f"Coefficients: {coefficients}")
print(f"Intercept: {intercept}")

Predicted size for sample [[750, 840]]: 1107682.186293602
Model R² Score: 0.8552
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.4795834075


/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


$h(x) = 866.1464 \cdot KCl + 1032.6951 \cdot Temperature - 409391.4796$

## Part 4. Use Cross Validation

In [21]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
cv_scores = cross_val_score(model, X, y, cv=5)
# Report on their finding and their significance
print("Cross-validation scores for each fold:", cv_scores)
print(f"Mean cross-validation score: {cv_scores.mean():.4f}")
print(f"Standard deviation of scores: {cv_scores.std():.4f}")
print("This is a significant cross-validation score with a high mean of close to 85.6 percent and a standard deviation of 1.3 percent.")
print("This gives us a pretty good interval to estimate the performance of our model.")

Cross-validation scores for each fold: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Mean cross-validation score: 0.8568
Standard deviation of scores: 0.0135
This is a significant cross-validation score with a high mean of close to 85.6 percent and a standard deviation of 1.3 percent.
This gives us a pretty good interval to estimate the performance of our model.


## Part 5. Using Polynomial Regression

In [22]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Report on the metrics and output the resultant equation as you did in Part 3.

poly = PolynomialFeatures(degree=2, include_bias=False)
model_poly = make_pipeline(poly, LinearRegression())
model_poly.fit(X_train, y_train)

poly_score = model_poly.score(X_test, y_test)
print(f"Polynomial Model R² Score: {poly_score:.4f}")

linear_model = model_poly.named_steps['linearregression']
coefficients = linear_model.coef_.flatten()
intercept = linear_model.intercept_
print(f"Coefficients: {coefficients}")
print(f"Intercept: {intercept}")

Polynomial Model R² Score: 1.0000
Coefficients: [ 1.20000000e+01 -1.27196688e-07  1.26183711e-11  2.00000000e+00
  2.85714287e-02]
Intercept: 2.0481296814978123e-05


$
h(x) = 12 \cdot KCl^2 - 1.27196688 \times 10^{-7} \cdot KCl \cdot Temperature + 
1.26183711 \times 10^{-11} \cdot Temperature^2 + 
2 \cdot KCl + 0.0285714287 \cdot Temperature + 2.0481296814978123 \times 10^{-5}
$
